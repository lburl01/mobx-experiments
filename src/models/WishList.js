import { types } from "mobx-state-tree";

const data = {
  name: "New Camera",
  price: 833.33,
  image:
    "https://images.unsplash.com/photo-1566864222010-d45675442c31?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1300&q=80"
};

export const WishListItem = types
  .model({
    name: types.string,
    price: types.number,
    image: ""
  })
  .actions(self => ({
    changeName(newName) {
      self.name = newName;
    },
    changePrice(newPrice) {
      self.price = newPrice;
    },
    changeImage(newImage) {
      self.image = newImage;
    }
  }));

export const WishList = types
  .model({
    items: types.optional(types.array(WishListItem), [])
  })
  .actions(self => ({
    add(item) {
      self.items.push(item);
    }
  }))
  .views(self => ({
    get totalPrice() {
      return self.items.reduce((sum, entry) => sum + entry.price, 0);
    }
  }));
