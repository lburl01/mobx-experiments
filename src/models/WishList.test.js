import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree";
import { reaction } from "mobx";

import { WishList, WishListItem } from "./WishList";

it("can create an instance of a model", () => {
  const item = WishListItem.create({
    name: "New Camera",
    price: 833.33
  });

  expect(item.price).toBe(833.33);
  expect(item.image).toBe("");
  item.changeName("Nikon Camera");
  expect(item.name).toBe("Nikon Camera");
});

it("can create a wishlist", () => {
  const list = WishList.create({
    items: [
      {
        name: "New Camera",
        price: 833.33
      }
    ]
  });
  expect(list.items.length).toBe(1);
  expect(list.items[0].price).toBe(833.33);
});

it("can add new items", () => {
  const list = WishList.create();
  const states = [];
  onSnapshot(list, snapshot => {
    states.push(snapshot);
  });

  list.add({
    name: "Waffle Maker",
    price: 25.0
  });

  expect(list.items.length).toBe(1);
  expect(list.items[0].name).toBe("Waffle Maker");
  list.items[0].changeName("Circus Waffle Maker");

  expect(getSnapshot(list)).toMatchSnapshot();
  expect(states).toMatchSnapshot();
});

it("can add new items with onPatch", () => {
  const list = WishList.create();
  const patches = [];
  onPatch(list, patch => {
    patches.push(patch);
  });

  list.add({
    name: "Waffle Maker",
    price: 25.0
  });

  list.items[0].changeName("Circus Waffle Maker");

  expect(patches).toMatchSnapshot();
});

it("can calculate the total price of a wishlist", () => {
  const list = WishList.create({
    items: [
      {
        name: "Waffle Maker",
        price: 25.0
      },
      {
        name: "New Camera",
        price: 833.33
      }
    ]
  });

  expect(list.totalPrice).toBe(858.33);

  let changed = 0;
  reaction(() => list.totalPrice, () => changed++);
  expect(changed).toBe(0);
  list.items[0].changeName("Test");
  expect(changed).toBe(0);
  list.items[0].changePrice(50);
  expect(changed).toBe(1);
});
