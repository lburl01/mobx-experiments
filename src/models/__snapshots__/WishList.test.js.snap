// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "Circus Waffle Maker",
      "price": 25,
    },
  ],
}
`;

exports[`can add new items 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Waffle Maker",
        "price": 25,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Circus Waffle Maker",
        "price": 25,
      },
    ],
  },
]
`;

exports[`can add new items with onPatch 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "Waffle Maker",
      "price": 25,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "Circus Waffle Maker",
  },
]
`;
